first step in the Oscarification of pc groups

ThomasBreuer opened the pull request:
- changed the definition of the `PcGroup` type from the GAP type `IsPcGroup` to the union of `IsPcGroup` and `IsPcpGroup` (which now allows for infinite groups),
- introduced some GAP filters for that,
- added `Collector` objects in Oscar from which the corresponding GAP objects are created on demand,
- changed the defaults for `cyclic_group`, `dihedral_group` to return `PcGroup`s also in the infinite case (no longer `FPGroup`s),
- support `abelian_group(PcGroup, ...)` for infinite groups,
- moved some documentation,
- small additions to make this work:
  - `free_group` now admits an optional keyword argument s.t. one can create a group with elements in syllable representation,
  - the syntax `G([1 =>2, 2 => -3])` is now supported for free groups `G`, in order to construct elements of `G`,
  - `PcGroup` does no longer imply finiteness.

ThomasBreuer committed:
first step in the Oscarification of pc groups

- changed the definition of the `PcGroup` type from the GAP type `IsPcGroup`
  to the union of `IsPcGroup` and `IsPcpGroup`
  (which now allows for infinite groups),
- introduced some GAP filters for that,
- added `Collector` objects in Oscar from which the corresponding
  GAP objects are created on demand,
- changed the defaults for `cyclic_group`, `dihedral_group` to return
  `PcGroup`s also in the infinite case (no longer `FPGroup`s),
- support `abelian_group(PcGroup, ...)` for infinite groups,
- moved some documentation,
- small additions to make this work:
  - `free_group` now admits an optional keyword argument s.t. one can create
    a group with elements in syllable representation,
  - the syntax `G([1 =>2, 2 => -3])` is now supported for free groups `G`,
    in order to construct elements of `G`,
  - `PcGroup` does no longer imply finiteness.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -151,7 +151,7 @@ false
 
-Base.isfinite(G::PcGroup) = true
+# Base.isfinite(G::PcGroup) = true
 
```
```suggestion
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1701,6 +1701,40 @@ function length(g::FPGroupElem)
+"""
+    (G::FPGroup)(l::AbstractVector{Pair{T, S}}) where {T <: IntegerUnion, S <: IntegerUnion}
+
+Return the element `x` of `G` that is described by `l`
```
The font used by my browser for GitHub render `l` quite similar to `i1j`. Perhaps we could use a slightly more distinctive name here, e.g. `list` or `pairs` or so

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1701,6 +1701,40 @@ function length(g::FPGroupElem)
+   for p in l
+     0 < p.first && p.first <= n || throw(ArgumentError("generator number is at most $n"))
+     push!(ll, p.first)
+     push!(ll, p.second)
```
`ll` is a `Vector{S}` but `first` is of type `T`. I guess in practice `T` and `S` both will be `Int` and this won't matter... But we could also sidestep all of this by changing `ll` into a `Vector{GAP.Obj}` and then doing this:
```suggestion
     push!(ll, GAP.Obj(p.first))
     push!(ll, GAP.Obj(p.second))
```
And then of course later we can just write `GapObj(ll)` w/o a need for recursive conversion.

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -360,7 +385,7 @@ quaternion_group(n::IntegerUnion) = quaternion_group(PcGroup, n)
   # FIXME: resolve naming: dicyclic vs (generalized) quaternion: only the
   # former should be for any n divisible by 4; the latter only for powers of 2.
-  # see also debate on the GAP side
+  # see also debate on the GAP side (gap-system/gap/issues/2725)
```
Good idea to add this reference, how about turning it into a full link?
```suggestion
  # see also debate on the GAP side (https://github.com/gap-system/gap/issues/2725)
```

fingolfin commented on src/Groups/pcgroup.jl:
```diff
@@ -0,0 +1,345 @@
+  GAP.evalstr("DeclareFilter(\"IsPcElementOrPcpElement\")");
+  GAP.evalstr("InstallTrueMethod(IsPcElementOrPcpElement, IsMultiplicativeElementWithInverseByPolycyclicCollector)");
+  GAP.evalstr("InstallTrueMethod(IsPcElementOrPcpElement, IsPcpElement)");
+  GAP.evalstr("BindGlobal(\"IsPcGroupOrPcpGroup\", IsGroup and CategoryCollections(IsPcElementOrPcpElement))");
```
Could also do tis (but do whatever you prefer, though)
```suggestion
  GAP.evalstr("""
      DeclareFilter("IsPcElementOrPcpElement");
      InstallTrueMethod(IsPcElementOrPcpElement, IsMultiplicativeElementWithInverseByPolycyclicCollector);
      InstallTrueMethod(IsPcElementOrPcpElement, IsPcpElement);
      BindGlobal("IsPcGroupOrPcpGroup", IsGroup and CategoryCollections(IsPcElementOrPcpElement));
      """)
```

fingolfin commented on src/Groups/pcgroup.jl:
```diff
@@ -0,0 +1,345 @@
+  if isdefined(c, :X)
+    if GAP.Globals.IsSingleCollectorRep(c.X)
+      G = c.F
+      GAP.Globals.SetConjugate(c.X, j, i, G(rhs).X)
```
perhaps we should also wrap `SetCommutator`?

fingolfin reviewed (commented):
Some quick remarks. I'd like to have a closer look after I've slept ;-)

fingolfin commented:
Has a conflict with `src/Oscar.jl`, hence CI does not run

fingolfin commented on src/Groups/pcgroup.jl:
```diff
@@ -0,0 +1,345 @@
+
+
+# Create the collector on the GAP side on demand
+function Base.getproperty(c::GAP_Collector, sym::Symbol)
```
We use `getproperty` like this for some of the other types mostly for legacy reasons, IMHO (namely because we did not provide a good way to access the underlying GAP object).

I think it would be preferable to change code which uses `collector.X` to get the underlying GAP object to use a helper method instead. The canonical idea would be to defined `GAPObj(collector)` for this role (i.e., what the code here does: compute `collector.X` if missing and then return it). Though we could also discuss other names, of course (`underlying_gap_object`?)

fingolfin commented on src/Groups/pcgroup.jl:
````diff
@@ -0,0 +1,345 @@
+```
+"""
+function pc_group(c::GAP_Collector)
+  if 0 in c.relorders
````
Could we perhaps instead check the filters of `c.X` to decide which way we go? Then in the future we could offer a way to optionally also create a pcp group even if the collector describes a finite group (I am not sure we *want* that, but at least it would be an option. Actually I can think of one advantage pcp groups have there: e.g. a cyclic group of order 10^6 can be described with one generator as a pcp group, but not as a pc group (at least not if one wants that pc group the useful)

fingolfin reviewed (changes requested):
Overall this is a good change and I approve, thank you very much! I left a few minor comments. Also, the merge conflict needs to be resolved

ThomasBreuer committed:
addressed the comments

The new function `set_commutator` delegates to `set_conjugate`,
analogous to the situation in GAP.
(In GAP, a delegation in the other direction is provided for collectors
in the filter `GAP.Globals.IsPowerCommutatorCollector`
but apparently all collectors supported by the GAP library and its
Polycyclic package are in `GAP.Globals.IsPowerConjugateCollector`.)

ThomasBreuer committed:
fixed a typo

ThomasBreuer committed:
fixed two tests

ThomasBreuer committed:
Merge branch 'master' into TB_pcgroup
